RavenDB-12898 Limit size of counter groups

aviv86 opened the pull request:
https://issues.hibernatingrhinos.com/issue/RavenDB-12898

ayende committed:
RavenDB-12898 - When counter group size exceed specified size, we'll now split it in half

aviv86 committed:
RavenDB-12898 : refactoring PutCounters() due to changes in counter-key format and counter-document splitting

aviv86 committed:
RavenDB-12898 : bug fixes

arekpalinski committed:
RavenDB-12931 Improving counters schema upgrade:

- splitting it into multiple transactions,
- making sure we're cleaning up pages occupied by legacy counter tables

aviv86 committed:
RavenDB-12898 : fix replication part

aviv86 committed:
RavenDB-12898 : CounterGroupDetail should use document id instead of CounterKey

aviv86 committed:
RavenDB-12898 : fix ETL part

aviv86 committed:
RavenDB-12898 : fix failing test (counter names are case-sensitive now)

aviv86 committed:
RavenDB-12898 : fix schema upgrade due to changes in counter-document splitting

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -502,45 +631,43 @@ private BlittableJsonReaderObject WriteNewCountersDocument(DocumentsOperationCon
-                            if (data.TryGet(Values, out BlittableJsonReaderObject localCounters) == false)
+                        if (table.SeekOnePrimaryKeyPrefix(documentKeyPrefix, out _) == false)
+                        {
+                            data = context.ReadObject(sourceData, documentId, BlittableJsonDocumentBuilder.UsageMode.ToDisk);
```
```suggestion
                            // simplest case of having no counters for this documents, can use the raw data from the source as-is
                            data = context.ReadObject(sourceData, documentId, BlittableJsonDocumentBuilder.UsageMode.ToDisk);
```

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -502,45 +631,43 @@ private BlittableJsonReaderObject WriteNewCountersDocument(DocumentsOperationCon
 
+                        else
+                        {
                             if (sourceData.TryGet(DbIds, out BlittableJsonReaderArray sourceDbIds) == false)
```
Let's add this validation _before_ the previous `if`. So we validate the structure before we save it as-is

aviv86 replied:
done

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -493,6 +565,63 @@ private BlittableJsonReaderObject WriteNewCountersDocument(DocumentsOperationCon
+            }
+        }
+
         public void PutCounters(DocumentsOperationContext context, string documentId, string collection, string changeVector,
```
This is pretty complex, and I think this requires additional code review (and probably some comments).
Let's sit on this tomorrow.

aviv86 committed:
RavenDB-12898 : fix PR comments

aviv86 committed:
RavenDB-12898 : PutCounters() - instead of inserting to table (and possibly doing context.ReadObject) on each iteration, keep 'entriesToUpdate' dictionary, so we will insert to table (and ReadObject if needed) only once at the end

aviv86 committed:
RavenDB-12898 : test that we can handle incoming counter replication when counter-group documents are spit differently

ayende commented on src/Raven.Server/Documents/CountersStorage.cs:
```diff
@@ -411,8 +414,14 @@ private static void SplitCounterGroup(DocumentsOperationContext context, Collect
-
+                    if (storageId.HasValue)
+                    {
+                        table.Update(storageId.Value, tvb);
```
Why do you need that?

aviv86 replied:
when I call `SplitCounterGroup` from `PutCounters`  I do not have the table value reader at hand anymore (because now I do that at the end, after the loop). 
when I call `SplitCounterGroup` from `IncrementCounter` I do have the value reader

ayende replied:
Yes, but why does this matter? Shouldn't we just always call `Set` ?

aviv86 committed:
RavenDB-12898 : refactoring

ayende committed:
RavenDB-12898
 - Added object pooling to reuse allocations
 - Minor refacotring

ayende committed:
RavenDB-12898 - Making the migration process work on batches and not create / split them

arekpalinski commented:
@ayende Regarding this commit: 1aa9d80b7da7d665d70e36dd1856836d5e5a638b. I'm currently rewriting `From16.cs` schema upgrade as it requires much more work - we need to split it into multiple transactions, we need to delete relevant tables, indexes etc

arekpalinski commented:
Oh, I see `From16.cs` contains more changes there. I'll rebase my work against this branch then.

aviv86 committed:
RavenDB-12898 : when we have a remote-delete + local-blob conflict that is resolved in favor of blob, we need to merge the deleted cv into blob

arekpalinski committed:
RavenDB-12898 Fixed paging calculations over counters batch

arekpalinski commented:
@aviv86 @ayende 
I've submitted the following PR https://github.com/aviv86/ravendb/pull/1 which splits the counters migration into multiple commits. Please review it there.

Also please have a look at the following comment: https://issues.hibernatingrhinos.com/issue/RavenDB-12898#focus=streamItem-67-35156.0-0

arekpalinski committed:
RavenDB-12931 PR Comment fixes
- no need to commit read tx
- checking if item is owned by table only if PK is global

arekpalinski committed:
RavenDB-12931 Checking if a node is fixed size tree header explicitly instead of relying on exception
